data/s3/src/main/java/org/teiid/spring/data/s3/S3Connection.java

```diff
@@ -0,0 +1,198 @@
+ */
+package org.teiid.spring.data.s3;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import org.teiid.file.VirtualFile;
+import org.teiid.file.VirtualFileConnection;
+import org.teiid.translator.TranslatorException;
+
+import java.io.InputStream;
+import java.util.Vector;
+
+public class S3Connection implements VirtualFileConnection {
+
+
+    private final S3Configuration s3Config;
+    private final AmazonS3 s3Client;
+
+    public S3Connection(S3Configuration s3Config, AmazonS3 s3Client) {
+        this.s3Config = s3Config;
+        this.s3Client = s3Client;
+    }
+
+    @Override
+    public VirtualFile[] getFiles(String s) throws TranslatorException {
+        if(s3Client.doesObjectExist(s3Config.getBucket(), s)){
+            ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
+                    .withBucketName(s3Config.getBucket())
+                    .withPrefix(s);
+            ObjectListing objectListing = s3Client.listObjects(listObjectsRequest);
+            return new VirtualFile[] {new S3VirtualFile(s3Client, objectListing.getObjectSummaries().get(0))};
+        }
+        if(isDirectory(s)){
+            if(!s.endsWith("/")) {
+                s += "/";
+            }
+            return convert(s);
+        }
+        String parentPath;
+        if(s.contains("/")) {
+            parentPath = s.substring(0, s.lastIndexOf('/'));
+        }
+        else {
+            parentPath = "";
+        }
+
+        if(!isDirectory(parentPath))
+            return null;
+        if(s.contains("*"))
+        {
+            return globSearch(parentPath, s);
+        }
+        return null;
+    }
+
+    private VirtualFile[] globSearch(String parentPath, String s) {
+        Vector<S3VirtualFile> s3VirtualFiles = new Vector<>();
+        ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
+                .withBucketName(s3Config.getBucket())
+                .withPrefix(parentPath)
+                .withDelimiter("/");
+        ObjectListing objectListing = s3Client.listObjects(listObjectsRequest);
+        while(objectListing != null) {
+            for(S3ObjectSummary s3ObjectSummary : objectListing.getObjectSummaries()) {
+                if(!s3ObjectSummary.getKey().endsWith("/")) {
+                    if(matchString(s3ObjectSummary.getKey(), s)) {
+                        s3VirtualFiles.add(new S3VirtualFile(s3Client, s3ObjectSummary));
+                    }
+                }
+            }
+            if(!objectListing.isTruncated())
+                break;
+            objectListing = s3Client.listNextBatchOfObjects(objectListing);
+        }
+        VirtualFile[] virtualFiles = new VirtualFile[s3VirtualFiles.size()];
```

Comment:
you can also use collection.asList(new VirtualFile[s3VirtualFiles.size()); to create the result array.

Reply:
Any java.util.Collection.  In this case s3VirtualFiles

Reply:
I don't think there's a method like asList() for an arrayList. What I think you are saying is to convert the array list to a array using the toArray() method but I tried it and didn't work 'cause this would return an array of the type S3VirtualFile which would eventually have to be referenced by a virtual file reference. I'm confused!!

Reply:
Sorry I mean toArray.   

s3VirtualFiles.toArray(new VirtualFile[s3VirtualFiles.size()]);

does the same thing as the code you have of manually constructing the array and setting each of the values from the list.

Reply:
It does. Thanks